modules/pubsub/src/main/java/com/spotify/ffwd/pubsub/PubsubPluginSink.java

```diff
@@ -82,52 +88,46 @@ public void init() { }
 
   @Override
   public AsyncFuture<Void> sendEvents(Collection<Event> events) {
-    log.debug("Sending events is not supported!");
+    logger.debug("Sending events is not supported!");
     return async.resolved();
   }
 
-  private void publishPubSub(ByteString bytes) {
-    final ApiFuture<String> publish = publisher.publish(
-      PubsubMessage.newBuilder().setData(bytes).build());
-
-    ApiFutures.addCallback(publish, new ApiFutureCallback<String>() {
-      @Override
-      public void onFailure(Throwable t) {
-        log.error("Failed sending metrics {}", t.getMessage());
-      }
-
-      @Override
-      public void onSuccess(String messageId) { }
-
-    }, executorService);
-  }
-
   @Override
   public AsyncFuture<Void> sendMetrics(Collection<Metric> metrics) {
     final UUID traceId = UUID.randomUUID();
-    log.info("{}: Start sending metrics", traceId);
+    logger.info("{}: Start sending metrics", traceId);
 
     for (Metric metric : metrics) {
       try {
-        final ApiFuture<String> publish = publisher.publish(PubsubMessage.newBuilder()
-          .setData(ByteString.copyFrom(serializer.serialize(metric)))
-          .build()
-        );
+        final ApiFuture<String> publish;
+        if (writeCache.isCached(metric)) {
+          publish = ApiFutures.immediateFuture("Metric in cache");
+        } else {
+          publish = publisher.publish(PubsubMessage.newBuilder()
+            .setData(ByteString.copyFrom(serializer.serialize(metric)))
+            .build()
+          );
+        }
+
         ApiFutures.addCallback(publish, new ApiFutureCallback<String>() {
```

Comment:
Since the async callbacks only are relevant to real publishes/cache misses, it would be easier to read if you add the publishPubSub function back and use that (with the cache changes of course). The loop in sendMetrics is starting to get large and messy.

Reply:
Oh the `publishPubSub` function wasn't being used, that's why I deleted it. That might have been left over from when we reverted the protobuf change(?). Anyway I can refactor this method.